File: generated-code/example/fsharp/Model/OneOf.fs
#nowarn "0058"
namespace TransGenTest.Model

type OneOfOptionOne = {
    VecInt: int[];
    LongInt: int64;
    } with
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 0
        writer.Write this.VecInt.Length
        this.VecInt |> Array.iter (fun value ->
            writer.Write value
        )
        writer.Write this.LongInt
    static member readFrom(reader: System.IO.BinaryReader) = {
        VecInt = [|for _ in 1 .. reader.ReadInt32() do
            yield reader.ReadInt32()
        |]
        LongInt = reader.ReadInt64()
    }

type OneOfOptionTwo = {
    Value: int;
    } with
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 1
        writer.Write this.Value
    static member readFrom(reader: System.IO.BinaryReader) = {
        Value = reader.ReadInt32()
    }
type OneOf = 
    | OptionOne of OneOfOptionOne
    | OptionTwo of OneOfOptionTwo
    with
    member this.writeTo(writer: System.IO.BinaryWriter) =
        match this with
            | OptionOne value -> value.writeTo writer
            | OptionTwo value -> value.writeTo writer
    static member readFrom(reader: System.IO.BinaryReader) =
        match reader.ReadInt32() with
            | 0 -> OptionOne (OneOfOptionOne.readFrom reader)
            | 1 -> OptionTwo (OneOfOptionTwo.readFrom reader)
            | x -> failwith (sprintf "Unexpected tag %d" x)

File: generated-code/example/csharp/Model/OneOf.cs
namespace TransGenTest.Model
{
    public abstract class OneOf
    {
        public abstract void WriteTo(System.IO.BinaryWriter writer);
        public static OneOf ReadFrom(System.IO.BinaryReader reader)
        {
            switch (reader.ReadInt32())
            {
                case OptionOne.TAG:
                    return OptionOne.ReadFrom(reader);
                case OptionTwo.TAG:
                    return OptionTwo.ReadFrom(reader);
                default:
                    throw new System.Exception("Unexpected tag value");
            }
        }

        public class OptionOne : OneOf
        {
            public const int TAG = 0;
            public int[] VecInt { get; set; }
            public long LongInt { get; set; }
            public OptionOne() {}
            public OptionOne(int[] vecInt, long longInt)
            {
                this.VecInt = vecInt;
                this.LongInt = longInt;
            }
            public static new OptionOne ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new OptionOne();
                result.VecInt = new int[reader.ReadInt32()];
                for (int i = 0; i < result.VecInt.Length; i++)
                {
                    result.VecInt[i] = reader.ReadInt32();
                }
                result.LongInt = reader.ReadInt64();
                return result;
            }
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                writer.Write(VecInt.Length);
                foreach (var VecIntElement in VecInt)
                {
                    writer.Write(VecIntElement);
                }
                writer.Write(LongInt);
            }
        }

        public class OptionTwo : OneOf
        {
            public const int TAG = 1;
            public int Value { get; set; }
            public OptionTwo() {}
            public OptionTwo(int value)
            {
                this.Value = value;
            }
            public static new OptionTwo ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new OptionTwo();
                result.Value = reader.ReadInt32();
                return result;
            }
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                writer.Write(Value);
            }
        }
    }
}
